src/main/java/net/eldeen/dropwizard/CfSignalResourceBundle.java

```diff
@@ -7,75 +17,154 @@
-   * @param awsRegion       the aws region the stack is in, e.g. us-west-2
-   */
-  public CfSignalResourceBundle(String asgResourceName, String stackName, String awsRegion) {
-    this.asgResourceName = asgResourceName;
-    this.stackName = stackName;
-    this.awsRegion = awsRegion;
+  private final Supplier<Optional<String>> instanceIdProvider;
+  private final Function<CfSignalResourceConfig, AmazonCloudFormation> cloudFormationSupplier;
+  private final AtomicReference<AmazonCloudFormation> cloudFormationAtomicRef = new AtomicReference<>(null);
+
+  public CfSignalResourceBundle() {
+    instanceIdProvider = () -> Optional.ofNullable(EC2MetadataUtils.getInstanceId());
+
+    cloudFormationSupplier = (cfSignalResourceConfig) -> {
+      AmazonCloudFormation amazonCloudFormation = cloudFormationAtomicRef.get();
+
+      if (amazonCloudFormation != null) {
+        return amazonCloudFormation;
+      }
+
+      return cloudFormationAtomicRef.updateAndGet((unused) -> {
+
+        AmazonCloudFormationClient amazonCloudFormationClient = new AmazonCloudFormationClient();
+
+        amazonCloudFormationClient.setRegion(Region.getRegion(Regions.valueOf(cfSignalResourceConfig.getAwsRegion()
+                                                                                                    .toUpperCase()
+                                                                                                    .replace('-', '_'))));
+        return amazonCloudFormationClient;
+      });
+    };
+  }
+
+  @Inject
+  public CfSignalResourceBundle(@CfSignalResourceInstanceId Optional<String> instanceId, AmazonCloudFormation amazonCloudFormation) {
+    instanceIdProvider = () -> checkNotNull(instanceId);
+    cloudFormationSupplier = (config) -> amazonCloudFormation;
   }
 
   @Override
   public void initialize(Bootstrap<?> bootstrap) {
+  }
 
+  public CfSignalResourceConfig getConfiguration() {
+    return new CfSignalResourceConfig();
   }
 
   @Override
-  public void run(Environment environment) {
+  public void run(T config, Environment environment) {
 
-    final String instanceId = EC2MetadataUtils.getInstanceId();
-    if (instanceId == null || instanceId.trim() == "") {
+    final Optional<String> instanceId = instanceIdProvider.get();
+    if (!instanceId.isPresent()) {
       LOGGER.debug("Unable to fetch EC2 Instance ID, assuming not running on AWS and thus not signalling");
       return;
     }
 
-    environment.lifecycle().addServerLifecycleListener(new ServerLifecycleListener() {
-      @Override
-      public void serverStarted(Server server) {
-        AmazonCloudFormation client = null;
+    environment.lifecycle()
+               .addLifeCycleListener(
+                 new CfSignalResourceLifcycleListener(getCfResourceBundleConfig(config).orElseGet(this::getConfiguration),
+                                                      instanceId.get()));
```

Comment:
Just read your response on the `getConfiguration` method. What if instead we make it throw e.g. `UnsupportedOperationException` by default?